controllers/secret_watch_controller.go

```diff
@@ -0,0 +1,240 @@
+			return err
+		}
+	}
+
+	return nil
+}
+
+// secretDataChanged checks if the secret's data actually changed (not just metadata)
+func (c *SecretWatchController) secretDataChanged(e event.UpdateEvent) bool {
+	oldSec, oldOk := e.ObjectOld.(*corev1.Secret)
+	newSec, newOk := e.ObjectNew.(*corev1.Secret)
+
+	if !oldOk || !newOk {
+		return false
+	}
+
+	return !reflect.DeepEqual(oldSec.Data, newSec.Data)
+}
+
+// findResourcesUsingSecret finds all resources that reference the given secret as connInfoSecretSource
+func (c *SecretWatchController) findResourcesUsingSecret(ctx context.Context, secret *corev1.Secret) ([]SecretSourceResource, error) {
+	secretKey := fmt.Sprintf("%s/%s", secret.Namespace, secret.Name)
+	var allResources []SecretSourceResource //nolint:prealloc
+
+	serviceUserList := &v1alpha1.ServiceUserList{}
+	err := c.List(ctx, serviceUserList, &client.ListOptions{
+		FieldSelector: fields.OneTermEqualSelector(connInfoSecretRefIndexKey, secretKey),
+	})
+	if err != nil {
+		return nil, fmt.Errorf("failed to list ServiceUsers: %w", err)
+	}
+	for i := range serviceUserList.Items {
+		allResources = append(allResources, &serviceUserList.Items[i])
+	}
+
+	clickhouseUserList := &v1alpha1.ClickhouseUserList{}
+	err = c.List(ctx, clickhouseUserList, &client.ListOptions{
+		FieldSelector: fields.OneTermEqualSelector(connInfoSecretRefIndexKey, secretKey),
+	})
+	if err != nil {
+		return nil, fmt.Errorf("failed to list ClickhouseUsers: %w", err)
+	}
+	for i := range clickhouseUserList.Items {
+		allResources = append(allResources, &clickhouseUserList.Items[i])
+	}
+
+	return allResources, nil
+}
+
+// triggerReconciliation triggers reconciliation by updating the resource's annotation
+func (c *SecretWatchController) triggerReconciliation(ctx context.Context, resource SecretSourceResource) error {
+	latest := resource.DeepCopyObject().(client.Object)
+	if err := c.Get(ctx, types.NamespacedName{
+		Name:      latest.GetName(),
+		Namespace: latest.GetNamespace(),
+	}, latest); err != nil {
+		return fmt.Errorf("failed to get latest version of resource: %w", err)
+	}
+
+	annotations := latest.GetAnnotations()
+	if annotations == nil {
+		annotations = make(map[string]string)
+	}
+	annotations["controllers.aiven.io/secret-source-updated"] = fmt.Sprintf("%d", time.Now().Unix())
+
+	// clear the processed generation annotation to force the basic controller to reconcile
+	delete(annotations, "controllers.aiven.io/generation-was-processed")
+
+	latest.SetAnnotations(annotations)
+
+	if err := c.Update(ctx, latest); err != nil {
+		return fmt.Errorf("failed to update resource annotation: %w", err)
+	}
+
+	c.Log.Info("triggered reconciliation for resource",
+		"resource", types.NamespacedName{Name: latest.GetName(), Namespace: latest.GetNamespace()},
+		"kind", latest.GetObjectKind().GroupVersionKind().Kind)
+
+	return nil
+}
```

Comment:
Please put the annotations into constants. They are shared with the testing code at least.

Reply:
done